contrib/hadoop-store-builder/src/java/voldemort/store/readonly/fetcher/HdfsDirectory.java

```diff
@@ -50,15 +55,52 @@ public HdfsFile getMetadataFile() {
         return metadataFile;
     }
 
-    public HdfsDirectory(FileSystem fs, Path source) throws IOException {
+    public HdfsDirectory(FileSystem fs, Path source, VoldemortConfig voldemortConfig) throws IOException {
         this.source = source;
 
-        FileStatus[] files = fs.listStatus(source);
+        FileStatus[] files = null;
+        int maxAttempts = voldemortConfig.getReadOnlyFetchRetryCount();
+        for (int attempt = 1; attempt <= maxAttempts && files == null; attempt++) {
+            try {
+                files = fs.listStatus(source);
+                break;
+            } catch (Exception e) {
+                if(attempt < maxAttempts) {
+                    // We may need to sleep
+                    long retryDelayMs = voldemortConfig.getReadOnlyFetchRetryDelayMs();
+                    if (retryDelayMs > 0) {
+                        // Doing random back off so that all nodes do not end up swarming the NameNode
+                        long randomDelay = (long) (Math.random() * retryDelayMs + retryDelayMs);
+
+                        logger.error("Could not execute listStatus operation on attempt # " + attempt +
+                                " / " + maxAttempts + ". Trying again in " + randomDelay + " ms.");
+                        try {
+                            Thread.sleep(randomDelay);
+                        } catch(InterruptedException ie) {
+                            logger.error("Fetcher interrupted while waiting to retry", ie);
+                            Thread.currentThread().interrupt();
+                        }
+                    }
+                } else {
+                    throw new VoldemortException("Failed " + maxAttempts +
+                            " times attempting listStatus operations for path: " + source, e);
+                }
+            }
+        }
+
         if(files == null) {
             throw new VoldemortApplicationException(source + " is empty");
         }
 
         for(FileStatus file: files) {
+            // Bookkeeping
+            totalSizeOfChildren += file.getLen();
```

Comment:
It should not be an issue for now , as we don't have subdirectories and the code path is invalid, we will clean that up later.